### PR TITLE
GPU enabled jupyter notebook for machine learning powered by Jupyter lab and Tensorflow

### PR DESCRIPTION
**Pull request by anuprulez**
This PR adds a new tool for Jupyter notebook powered by Jupyter lab to prototype and create machine learning models. The notebook (its docker container- https://hub.docker.com/repository/docker/anupkumar/docker-ml-jupyterlab (tag: galaxy-integration)) is built on top of [jupyter/tensorflow](https://hub.docker.com/r/jupyter/tensorflow-notebook) and adds GPU support for parallel computation. The tool is still in the development phase and there are more features such as model sharing that still need to be added. To run it in local Galaxy instance, two config files should be updated to run interactive tools - `galaxy.yml` and `job_conf.xml` The corresponding changes to be made in these files are mentioned in their respective `.interactivetools` files (`galaxy.yml.interactivetools` and `job_conf.xml.interactivetools`). In the `job_conf.xml` file, one extra setting 

`<param id="docker_run_extra_arguments">&#45;&#45;gpus all</param>` 

should be added for GPU computation. In addition, the newly added tool should be added to `tool_conf.xml` file.

notebook_jupyter_lab

gpu_usage

Please feel free to share your comments. Thank you!

Edit:

<param id="docker_run_extra_arguments">&#45;&#45;gpus all</param> setting in the job_conf.xml file should only be used when the docker command is executing on a VM with GPUs. Otherwise, this option should be excluded when the backend only has CPUs to avoid error such as
docker: Error response from daemon: could not select device driver "" with capabilities: [[gpu]].

### CODE REVIEW

File: tools/interactive/interactivetool_ml_jupyter_notebook.xml [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+                <param name="ipynb" type="data" format="ipynb" label="IPython Notebook"/>
+                <param name="run_it" type="boolean" truevalue="true" falsevalue="false" label="Execute notebook and return a new one."
+                    help="This option is useful in workflows when you just want to execute a notebook and not dive into the webfrontend."/>
+            </when>
+        </conditional>
+        <param name="input" type="data" optional="true" label="Include data into the environment"/>
+
+    </inputs>
+    <outputs>
+        <data name="jupyter_notebook" format="ipynb" label="Executed Notebook"></data>
```
Comment (nsoranzo) [inline tools/interactive/interactivetool_ml_jupyter_notebook.xml]:
```suggestion
        <data name="jupyter_notebook" format="ipynb" label="Executed Notebook" />
```

File: tools/interactive/interactivetool_ml_jupyter_notebook.xml [hunk 2]
```diff
@@ -0,0 +1,94 @@
+<tool id="interactive_tool_ml_jupyter_notebook" tool_type="interactive" name="GPU enabled Interactive Jupyter Notebook for Machine Learning" version="0.1">
+    <requirements>
+        <container type="docker">docker.io/anupkumar/docker-ml-jupyterlab:galaxy-integration</container>
+    </requirements>
+    <entry_points>
+        <entry_point name="GPU enabled Interactive Jupyter Notebook for Machine Learning" requires_domain="True">
+            <port>8888</port>
+	    <url>ipython/lab</url>
```
Comment (nsoranzo) [inline tools/interactive/interactivetool_ml_jupyter_notebook.xml]: Can you replace tabs with spaces here and below please?

File: tools/interactive/interactivetool_ml_jupyter_notebook.xml [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            <param name="mode_select" type="select" label="Do you already have a notebook?" help="If not, no problem we will provide you with a default one.">
+                <option value="scratch">Start with a fresh notebook</option>
+                <option value="previous">Load a previous notebook</option>
+            </param>
+            <when value="scratch"/>
+            <when value="previous">
+                <param name="ipynb" type="data" format="ipynb" label="IPython Notebook"/>
+                <param name="run_it" type="boolean" truevalue="true" falsevalue="false" label="Execute notebook and return a new one."
+                    help="This option is useful in workflows when you just want to execute a notebook and not dive into the webfrontend."/>
+            </when>
```
Comment (nsoranzo) [inline tools/interactive/interactivetool_ml_jupyter_notebook.xml:77]: The parameters from this branch of the conditional seem unused in the `<command>`.

File: tools/interactive/interactivetool_ml_jupyter_notebook.xml [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    </outputs>
+    <tests>
+        <test expect_num_outputs="1">
+            <param name="mode" value="previous" />
+            <param name="ipynb" value="test.ipynb" />
+            <param name="run_it" value="true" />
+            <output name="jupyter_notebook" file="test.ipynb" ftype="ipynb"/>
+        </test>
+    </tests>
+    <help>
```
Comment (nsoranzo) [inline tools/interactive/interactivetool_ml_jupyter_notebook.xml]:
```suggestion
    <help><![CDATA[
```

File: tools/interactive/interactivetool_ml_jupyter_notebook.xml [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+    JupyterLab is a next-generation web-based user interface for Project Jupyter. JupyterLab enables you to work with documents and activities such as Jupyter notebooks, text editors, 
+    terminals, and custom components in a flexible, integrated, and extensible manner.
+
+    Galaxy offers you to use Jupyter Lab directly in Galaxy accessing and interacting with Galaxy datasets as you like. A very common use-case is to
+    do the heavy lifting such as performing computation on GPUs and data reduction steps in Galaxy and the plotting and more `interactive` part on smaller datasets in Jupyter Lab.
+
+    You can start with a new Jupyter Lab notebook from scratch and wait until the job starts running. Running job will provide you a link which can be opened in the same or another 
+    browser tab. This link opens Jupyter Lab notebook which can be used to prototype machine learning solutions.
+
+    </help>
```
Comment (nsoranzo) [inline tools/interactive/interactivetool_ml_jupyter_notebook.xml]:
```suggestion
    ]]></help>
```

File: tools/interactive/interactivetool_ml_jupyter_notebook.xml [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+            chown \$NB_USER ./home_page.ipynb &&
+            chown \$NB_USER ./notebooks/ &&
+            chown \$NB_USER ./notebooks/*.ipynb &&
+            chown \$NB_USER ./elyra/ &&
+            chown \$NB_USER ./elyra/*.* &&
+            jupyter trust ./home_page.ipynb &&
+            jupyter trust ./notebooks/*.ipynb &&
+            jupyter trust ./elyra/*.ipynb &&
+            jupyter lab --allow-root --no-browser --NotebookApp.shutdown_button=True
+	#end if
```
Comment (nsoranzo) [inline tools/interactive/interactivetool_ml_jupyter_notebook.xml]:
```suggestion
        #end if
```

File: tools/interactive/interactivetool_ml_jupyter_notebook.xml [hunk 7]
```diff
@@ -0,0 +1,94 @@
+<tool id="interactive_tool_ml_jupyter_notebook" tool_type="interactive" name="GPU enabled Interactive Jupyter Notebook for Machine Learning" version="0.1">
+    <requirements>
+        <container type="docker">docker.io/anupkumar/docker-ml-jupyterlab:galaxy-integration</container>
```
Comment (mvdbeek) [inline tools/interactive/interactivetool_ml_jupyter_notebook.xml]: Can you include a version for the container ?

File: tools/interactive/interactivetool_ml_jupyter_notebook.xml [hunk 8]
```diff
@@ -0,0 +1,94 @@
+<tool id="interactive_tool_ml_jupyter_notebook" tool_type="interactive" name="GPU enabled Interactive Jupyter Notebook for Machine Learning" version="0.1">
```
Comment (mvdbeek) [inline tools/interactive/interactivetool_ml_jupyter_notebook.xml]:
```suggestion
<tool id="interactive_tool_ml_jupyter_notebook" tool_type="interactive" name="GPU enabled Interactive Jupyter Notebook for Machine Learning" version="0.1" profile="20.09">
```

File: tools/interactive/interactivetool_ml_jupyter_notebook.xml (hunk 7 above)
Comment (anuprulez) [reply tools/interactive/interactivetool_ml_jupyter_notebook.xml]: @mvdbeek Thanks for your comment, but I am not sure what should be the version number of a container when the tag name of the container is present. Can you give an example of a version? Thanks!

Comment (mvdbeek) [reply tools/interactive/interactivetool_ml_jupyter_notebook.xml]: You can add a version at the end of the tag?

Comment (anuprulez) [reply tools/interactive/interactivetool_ml_jupyter_notebook.xml]: ok, something like `:tag-name-version`. 
An example could be - `docker.io/anupkumar/docker-ml-jupyterlab:galaxy-integration-0.1`?

Comment (mvdbeek) [reply tools/interactive/interactivetool_ml_jupyter_notebook.xml]: sounds good

File: tools/interactive/interactivetool_ml_jupyter_notebook.xml (hunk 3 above)
Comment (mvdbeek) [reply tools/interactive/interactivetool_ml_jupyter_notebook.xml:77]: They are used in the else branch

**Review by mvdbeek — APPROVED** [review]

**Comment by anuprulez** [comment]
Thanks a lot @mvdbeek

File: tools/interactive/interactivetool_ml_jupyter_notebook.xml (hunk 3 above)
Comment (nsoranzo) [reply tools/interactive/interactivetool_ml_jupyter_notebook.xml:77]: > They are used in the else branch

Yep, after my review :wink:

**Pull request merged**